pkg/network/usm/utils/connection_rollup.go

```diff
@@ -210,34 +188,54 @@ func (c *ConnectionAggregator) ClearEphemeralPort(key types.ConnectionKey) types
 		return key
 	}
 
-	normalizedKey, _ := normalizeByIP(key)
-	ips, ports := split(normalizedKey)
-	savedPorts, ok := c.data[ips]
-	if !ok || savedPorts.ephemeralSide == ephemeralPortUnknown ||
-		(ports.a != savedPorts.a && ports.b != savedPorts.b) {
+	ips, ports, flipped := splitKey(key)
+	savedPorts, ok := c.findSimilar(ips, ports)
+	if !ok || savedPorts.serverSide == serverPortUnknown {
 		// We either haven't seen at this connection, or were not able to
 		// determine the ephemeral port side. In this case we return the key
 		// completely unmodified.
 		return key
 	}
 
-	// Get the server port from our stored information
-	serverPort := savedPorts.a
-	if savedPorts.ephemeralSide == ephemeralPortA {
-		serverPort = savedPorts.b
+	// We were able to determine the server side, in which case we clear the
+	// ephemeral port side and return the new key.
+	if savedPorts.serverSide == serverPortA {
+		savedPorts.b = 0
+	} else {
+		savedPorts.a = 0
 	}
 
-	// Clear the ephemeral port side
-	if key.DstPort == serverPort {
-		key.SrcPort = 0
-	} else {
-		key.DstPort = 0
+	return generateKey(ips, *savedPorts, flipped)
+}
+
+func (c *ConnectionAggregator) findSimilar(ips ipTuple, ports portValues) (*portValues, bool) {
+	pvSamples, ok := c.data[ips]
+	if !ok {
+		// a `find` with no matches is always followed by a insert by the code
+		// upstream so we go ahead a pre-allocate a slice for this ipTuple key
+		c.data[ips] = make([]portValues, 0, defaultServerPortSize)
+		return nil, false
 	}
 
-	return key
+	// NOTE: we're doing a brute force search here because the search space
```

Comment:
You can verify it using the profiler in the load test environment and once it's deployed to staging/production